snafu-derive/src/lib.rs

```diff
@@ -674,6 +674,57 @@ trait GenericAwareNames {
         Box::new(quote! { #enum_name<#(#original_generics,)*> })
     }
 
+    fn provided_generic_types_without_defaults(&self) -> Vec<proc_macro2::TokenStream> {
+        use syn::TypeParam;
+        self.generics()
+            .type_params()
+            .map(|t: &TypeParam| {
+                let &TypeParam {
+                    ref attrs,
+                    ref ident,
+                    ref colon_token,
+                    ref bounds,
+                    ..
+                } = t;
+                quote! {
+                    #(#attrs)*
+                    #ident
+                    #colon_token
+                    #bounds
+                }
+            })
+            .collect()
+    }
+
+    fn provided_generic_types_and_consts(&self) -> Vec<proc_macro2::TokenStream> {
+        use syn::{ConstParam, GenericParam, TypeParam};
+
+        self.generics()
+            .params
+            .iter()
+            .filter_map(|p| match *p {
+                GenericParam::Type(TypeParam { ref ident, .. }) => Some(quote! { #ident }),
+                GenericParam::Const(ConstParam { ref ident, .. }) => Some(quote! { #ident }),
+                _ => None,
+            })
+            .collect()
+    }
+
+    fn provided_generic_lifetimes(&self) -> Vec<proc_macro2::TokenStream> {
+        use syn::{GenericParam, LifetimeDef};
+
+        self.generics()
+            .params
+            .iter()
+            .filter_map(|p| match *p {
```

Comment:
I prefer to use `flat_map` instead of `filter_map`, since the former is strictly more powerful than the latter. I'd appreciate it if you could change these calls.